types/rdf-dataset-indexed/dataset.d.ts

```diff
@@ -0,0 +1,34 @@
+import { DatasetCore, DatasetFactory, DataFactory, Quad, BaseQuad, Stream, Term } from 'rdf-js';
+import { Readable } from 'stream';
+
+declare namespace DatasetIndexed {
+    interface DatasetIndexed<Q extends BaseQuad = Quad, InQuad extends BaseQuad = Quad> extends DatasetCore<Q, InQuad> {
+        create(quads: DatasetIndexed<BaseQuad, BaseQuad> | BaseQuad[]): this;
+        readonly length: number;
+        addAll(quads: DatasetIndexed<BaseQuad, BaseQuad> | BaseQuad[]): this;
+        clone(): this;
+        difference(other: this): this;
+        every(predicate: (quad: Q) => boolean): boolean;
+        filter(predicate: (quad: Q) => boolean): this;
+        forEach(callback: (quad: Q) => void): void;
+        import(stream: Stream): Promise<this>;
+        includes(quad: Quad): boolean;
+        intersection(other: this): this;
+        map(callback: (quad: Q) => Q): this;
+        merge(other: DatasetIndexed<BaseQuad, BaseQuad> | BaseQuad[]): this;
+        remove(quad: BaseQuad): this;
+        removeMatches(subject?: Term | null, predicate?: Term | null, object?: Term | null, graph?: Term | null): this;
+        some(predicate: (quad: Q) => boolean): boolean;
+        toArray(): Q[];
+        toStream(): Stream<Q> & Readable;
+    }
+}
+
+interface DatasetIndexed<Q extends BaseQuad = Quad, InQuad extends BaseQuad = Quad> extends DatasetIndexed.DatasetIndexed<Q> {}
+
+// tslint:disable-next-line no-unnecessary-class
+declare class DatasetIndexed<Q> {
```

Comment:
This is complicated. Any reason not to have the methods directly on the class, and implement `DatasetCore` here?

Reply:
The reason is maintenance and DRYness.

If you repeat all the methods, any time `DatasetCore` changes any implementing class has to be updated. And it can happen that Typescript will not signal that, causing inconsistency.

This way only the actually overridden or additional members are declared

Reply:
> If you repeat all the methods, any time `DatasetCore` changes any implementing class has to be updated.

Guess this refers to the lack of semver on `@types/*`? Agree that is a problem...

Reply:
Not necessarily. In the end this is just useless repetition...